### ISSUE TITLE
Uncaught Error: Failed to compile fragment shader.

### ISSUE THREAD

**Issue by PiotrMi** [PAIR-code/deeplearnjs#251]
When trying out the "Model Builder" demo in GPU mode it doesn´t work and this is logged in the console:

```
bundle.js:11668 1
2      precision highp float;
3      precision highp int;
4      varying vec2 resultUV;
5      const vec2 halfCR = vec2(0.5, 0.5);
6
7      bool isNaN(float val) {
8        return val == val ? false : true;
9      }
10
11     bool hasNaN(vec4 values) {
12       return any(notEqual(values, values));
13     }
14
15     float getNaN(vec4 values) {
16       return dot(vec4(1), values);
17     }
18
19     int round(float value) {
20       return int(floor(value + 0.5));
21     }
22
23     int imod(int x, int y) {
24       return x - y * (x / y);
25     }
26
27     const vec2 randomConst = vec2(
28       23.14069263277926, // e^pi (Gelfond's constant)
29        2.665144142690225 // 2^sqrt(2) (Gelfond–Schneider constant)
30     );
31
32     float random(float seed) {
33         return fract(cos(dot(resultUV * seed, randomConst)) * 12345.6789);
34     }
35
36     float sampleUVAndDepth(sampler2D texture, vec2 uv, int depth) {
37       float value;
38       if (depth == 0) {
39         value = texture2D(texture, uv).r;
40       } else if (depth == 1) {
41         value = texture2D(texture, uv).g;
42       } else if (depth == 2) {
43         value = texture2D(texture, uv).b;
44       } else if (depth == 3) {
45         value = texture2D(texture, uv).a;
46       }
47       return floor(value * 255.0 + 0.5);
48     }
49
50
51   vec2 UVfrom1D(int texNumR, int texNumC, int index) {
52     int texR = index / texNumC;
53     int texC = index - texR * texNumC;
54     return (vec2(texC, texR) + halfCR) / vec2(texNumC, texNumR);
55   }
56
57
58   vec2 UVfrom2D(int texNumR, int texNumC, int numC, int row, int col) {
59     int index = row * numC + col;
60     int texR = index / texNumC;
61     int texC = index - texR * texNumC;
62     return (vec2(texC, texR) + halfCR) / vec2(texNumC, texNumR);
63   }
64
65
66   vec2 UVfrom3D(int texNumR, int texNumC, int stride0,
67       int stride1, int row, int col, int depth) {
68     // Explicitly use integer operations as dot() only works on floats.
69     int index = row * stride0 + col * stride1 + depth;
70     int texR = index / texNumC;
71     int texC = index - texR * texNumC;
72     return (vec2(texC, texR) + halfCR) / vec2(texNumC, texNumR);
73   }
74
75
76   vec2 UVfrom4D(int texNumR, int texNumC, int stride0,
77       int stride1, int stride2, int row, int col, int depth,
78       int depth2) {
79     // Explicitly use integer operations as dot() only works on floats.
80     int index = row * stride0 + col * stride1 + depth * stride2 + depth2;
81     int texR = index / texNumC;
82     int texC = index - texR * texNumC;
83     return (vec2(texC, texR) + halfCR) / vec2(texNumC, texNumR);
84   }
85
86
87
88
89     float sample(sampler2D texture, vec2 uv) {
90       return texture2D(texture, uv).r;
91     }
92
93
94     void setOutput(float val) {
95       gl_FragColor = vec4(val, 0, 0, 0);
96     }
97
98   uniform sampler2D A;
99   uniform sampler2D B;
100
101
102      float getA() {
103        return sample(A, halfCR);
104      }
105
106      float getAAtOutCoords() {
bundle.js:11669 ERROR: 0:107: 'getOutputCoords' : no matching overloaded function found
bundle.js:11670  107        int coords = getOutputCoords();
bundle.js:11671 108
109        return getA();
110      }
111
112        float getAFlat(int index) {
113          return sample(A, halfCR);
114        }
115
116
117      float getB() {
118        return sample(B, halfCR);
119      }
120
121      float getBAtOutCoords() {
122        int coords = getOutputCoords();
123
124        return getB();
125      }
126
127        float getBFlat(int index) {
128          return sample(B, halfCR);
129        }
130
131
132        float binaryOperation(float a, float b) {
133          return a + b;
134        }
135
136        void main() {
137          float a = getAAtOutCoords();
138          float b = getBAtOutCoords();
139          setOutput(binaryOperation(a, b));
140        }
141
bundle.js:11644 Uncaught Error: Failed to compile fragment shader.
    at Object.createFragmentShader (bundle.js:11644)
    at GPGPUContext.createProgram (bundle.js:9982)
    at Object.compileProgram (bundle.js:10197)
    at bundle.js:8584
    at NDArrayMathGPU.getAndSaveBinary (bundle.js:8792)
    at NDArrayMathGPU.compileAndRun (bundle.js:8583)
    at NDArrayMathGPU.addInternal (bundle.js:8653)
    at bundle.js:7172
    at NDArrayMathGPU.NDArrayMath.executeOp (bundle.js:6926)
    at NDArrayMathGPU.NDArrayMath.add (bundle.js:7172)
createFragmentShader @ bundle.js:11644
GPGPUContext.createProgram @ bundle.js:9982
compileProgram @ bundle.js:10197
(anonymous) @ bundle.js:8584
NDArrayMathGPU.getAndSaveBinary @ bundle.js:8792
NDArrayMathGPU.compileAndRun @ bundle.js:8583
NDArrayMathGPU.addInternal @ bundle.js:8653
(anonymous) @ bundle.js:7172
NDArrayMath.executeOp @ bundle.js:6926
NDArrayMath.add @ bundle.js:7172
Session.updateCostForExample @ bundle.js:5801
(anonymous) @ bundle.js:5792
NDArrayMath.scope @ bundle.js:6807
Session.train @ bundle.js:5781
(anonymous) @ bundle.js:6145
NDArrayMath.scope @ bundle.js:6807
GraphRunner.trainNetwork @ bundle.js:6144
GraphRunner.train @ bundle.js:6113
ModelBuilder.startTraining @ bundle.js:598
(anonymous) @ bundle.js:469

```

Changing the dataset didn´t seem to help and when using CPU mode it became extremely slow (I used it a few weeks ago and CPU mode was way slower than GPU but the webpage was still a lot more responsive). 
I already used the demo a few weeks ago in GPU and CPU mode and it worked fine.

(Not sure if relevant) System specs:
Windows 7
Nvidia Geforce GTX 770

**Comment by dsmilkov** [comment]
Thanks for surfacing this! This bug will be fixed as soon as #248 is committed. The fix is a side-effect of optimizing the compiler to do more efficient batching & broadcasting.

**Comment by dsmilkov** [comment]
Fixed in #248
